### PR TITLE
Use mongo bulk updates for update requests

### PR DESCRIPTION
**Pull request by BVulaj**

### CODE REVIEW

**Comment by bserdar** [comment]
Please add the logic below to your PR as well:

https://github.com/lightblue-platform/lightblue-mongo/pull/282

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java [hunk 1]
```diff
@@ -53,7 +53,7 @@
 
     private static final Logger LOGGER = LoggerFactory.getLogger(BasicDocSaver.class);
 
-    private int batchSize = 64;
+    private final int batchSize = 64;
```
Comment (dcrissman) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java]: Should this be configurable?

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java [hunk 2]
```diff
@@ -48,28 +59,39 @@
 
     private static final Logger LOGGER = LoggerFactory.getLogger(IterateAndUpdate.class);
 
+    private final int batchSize = 64;
```
Comment (dcrissman) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: I see this declared a second time. Maybe it should be defined in a single location?

Comment (BVulaj) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: Three times, actually :(

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java (hunk 1 above)
Comment (BVulaj) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java]: It's definitely a configuration candidate, but @bserdar and I just decided to leave it here for now and move it out in another PR.  Is there a place in LB today to wire up configuration properties?

Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java]: Lets leave it in there for now. There is a ticket to move that to
configuration.

Comment (dcrissman) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java]: Probably the datasources.json, but I would be fine if we spawned a ticket to address this later as a configuration.

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java (hunk 2 above)
Comment (dcrissman) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: MongoMetadataConstants maybe? At least then we wouldn't loose them later when we do go to make it configurable.

Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: How about we get the batchsize in constructor now, declare a constant in
mongocrudcontroller and pass that? Later we can get that value from the
configuration.

Comment (BVulaj) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: I like it.  Incoming.

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java [hunk 3]
```diff
@@ -160,14 +192,63 @@ public void update(CRUDOperationContext ctx,
                 }
                 docIndex++;
             }
+            // if we have any remaining items to update
+            if (numUpdating > 0) {
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: This is never zero. numUpdating always increases.

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java [hunk 4]
```diff
@@ -160,14 +192,63 @@ public void update(CRUDOperationContext ctx,
...
-        response.setNumMatched(docIndex);
+        response.setNumMatched(numMatched);
+    }
+
+    private void handleBulkWriteError(List<BulkWriteError> errors, List<DocCtx> docs) {
+        docUpdateActual.addAll(docs);
+        for (BulkWriteError e : errors) {
+            DocCtx doc = docs.get(e.getIndex());
+            // remove any docs from the updated list if they had an error
+            docUpdateActual.set(e.getIndex(), null);
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: ?

Comment (BVulaj) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: Had initially needed to remove items from the list without affecting indexes, but removed that code before committing.  Will just remove this.

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java (hunk 3 above)
Comment (BVulaj) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: Resetting it in the above block now.

**Comment by bserdar** [comment]
Can you add one more unit test where an updated doc in the middle of the list fails. You create a unique index on something, and then create a duplicate record with an update to make it fail.

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java [hunk 5]
```diff
@@ -160,14 +194,62 @@ public void update(CRUDOperationContext ctx,
...
+        docUpdateActual.addAll(docs);
+        for (BulkWriteError e : errors) {
+            DocCtx doc = docs.get(e.getIndex());
+            if (e.getCode() == 11000 || e.getCode() == 11001) {
+                doc.addError(Error.get("update", MongoCrudConstants.ERR_DUPLICATE, e.getMessage()));
+            } else {
+                doc.addError(Error.get("update", MongoCrudConstants.ERR_SAVE_ERROR, e.getMessage()));
+            }
+        }
+        numFailed += (int) errors.stream().map(e -> e.getIndex()).distinct().count();
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: Instead of this, inc numFailed in the loop above.

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java [hunk 6]
```diff
@@ -160,14 +194,62 @@ public void update(CRUDOperationContext ctx,
...
+        });
+
         response.setNumUpdated(numUpdated);
         response.setNumFailed(numFailed);
-        response.setNumMatched(docIndex);
+        response.setNumMatched(numMatched);
+    }
+
+    private void handleBulkWriteError(List<BulkWriteError> errors, List<DocCtx> docs) {
+        docUpdateActual.addAll(docs);
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: This has all the docs, not just the actual updated ones.

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java [hunk 7]
```diff
@@ -160,14 +194,62 @@ public void update(CRUDOperationContext ctx,
...
+        for (BulkWriteError e : errors) {
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: You don't really need docupdateActual, ctx has methods that'll give the the list of docs with no errors.

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java [hunk 8]
```diff
@@ -160,14 +189,60 @@ public void update(CRUDOperationContext ctx,
...
+                LOGGER.debug("Bulk deleting docs with writeConcern={} from execution", writeConcern);
+                result = bwo.execute(writeConcern);
+            }
+        } catch (BulkWriteException e) {
+            BulkWriteException bwe = e;
+            result = bwe.getWriteResult();
+            LOGGER.warn("Bulk update operation contains errors", bwe);
+            docUpdateErrors.addAll(bwe.getWriteErrors());
+        } catch (Exception e) {
+            throw e;
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: Either log it, or don't catch it

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java [hunk 9]
```diff
@@ -160,14 +189,60 @@ public void update(CRUDOperationContext ctx,
                 }
                 docIndex++;
             }
+            // if we have any remaining items to update
+            if (numUpdating > 0) {
+                try {
+                    executeAndLogBulkErrors(bwo);
+                    numFailed += (int) docUpdateErrors.stream().map(e -> e.getIndex()).distinct().count();
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: I don't like this. Why not set this it in handleBulkWriteError

Comment (BVulaj) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: This whole block?  Or just the numFailed?  It was there previously but a previous comment asked for me to move it.  I could have misunderstood.

Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: numFailed. Can simply do numField++ every time you process an error

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java [hunk 10]
```diff
@@ -160,14 +189,58 @@ public void update(CRUDOperationContext ctx,
...
             }
         }
-        response.setNumUpdated(numUpdated);
+        handleBulkWriteError(docUpdateErrors, docUpdateAttempts);
+
+        ctx.getDocumentsWithoutErrors().stream().forEach(doc -> {
+            ctx.getFactory().getInterceptors().callInterceptors(InterceptPoint.POST_CRUD_UPDATE_DOC, ctx, doc);
+        });
+
+        response.setNumUpdated(ctx.getDocumentsWithoutErrors().size());
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: getDocumentsWithoutErrors is not numUpdated, it is exactly what it says: docsWithoutErrors, and it is an expensive operation. Instead, set a variable where you process errors.

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java (hunk 9 above)
Comment (BVulaj) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: The only reason I do this is because it seems possible for > 1 error to be returned per index, which would skew the actual number failed.  Do you think we should just not worry about that case?

Comment (BVulaj) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: If we want to ignore that case, we can simply just use the size of the error and update attempt lists to set numUpdated and numFailed

Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/IterateAndUpdate.java]: A better way of doing this is to concentrate the logic to a single place:
there is already a function that determines what's an error and what's not.
Count the failed docs there.
